#!/bin/bash
# usage: sbatch --nodes=1 --partition=sixhour "/panfs/pfs.local/home/j926w878/work/conda/snakecap/test13.sh" "A" "/panfs/pfs.local/home/j926w878/scratch/scratch_v3/SequenceCapture/SnakeCap_AllSamples/blast_database10/" "/panfs/pfs.local/home/j926w878/scratch/scratch_v3/SequenceCapture/SnakeCap_AllSamples/blast_database10/groupMatrix_consensusSeqCapAndTargetRef_vs_genomes.txt" "/panfs/pfs.local/home/j926w878/scratch/scratch_v3/SequenceCapture/SnakeCap_AllSamples/blast_database10/edgeMatrix_consensusSeqCapAndTargetRef_vs_genomes.txt" "/panfs/pfs.local/home/j926w878/scratch/scratch_v3/SequenceCapture/SnakeCap_AllSamples/blast_database10/groupAlignments_SeqCapAndTargetRef_consensus_InputSeqsMax100.fasta" # "NA"
# module load R
module load anaconda
conda activate py36

### Description:
# For each supergroup, creates a FASTA file containing the sequences assigned to the supergroup from (1) sample genomes and (2) consensus sequence of each SeqCap and TargRef group alignments
### NOTE: run this if you cause an infinite loop:  squeue -u $USER -h | awk '{print $1}' | xargs scancel

### Captured variables
MODE=$1  ### "A" for A/B; "C" for C/D; MODE C should be run after mode A/B to 

echo $MODE

BLASTDBDIR=$2
SUPERGROUPMATPATH=$3                          # path to the supergroup matrix generated by test12.sh
EDGEMATPATH=$4                                # path to the edge matrix generated by test12.sh
CONSENSUSPATH=$5                              # path to the fasta file containing consensus sequences of the alignment of SeqCap and TargRef sequences
#VARi=$6 #

# BLASTDBDIR="/panfs/pfs.local/home/j926w878/scratch/scratch_v3/SequenceCapture/SnakeCap_AllSamples/blast_database10/"
# SUPERGROUPMATPATH="/panfs/pfs.local/home/j926w878/scratch/scratch_v3/SequenceCapture/SnakeCap_AllSamples/blast_database10/groupMatrix_consensusSeqCapAndTargetRef_vs_genomes.txt"
# EDGEMATPATH="/panfs/pfs.local/home/j926w878/scratch/scratch_v3/SequenceCapture/SnakeCap_AllSamples/blast_database10/edgeMatrix_consensusSeqCapAndTargetRef_vs_genomes.txt"
# CONSENSUSPATH="/panfs/pfs.local/home/j926w878/scratch/scratch_v3/SequenceCapture/SnakeCap_AllSamples/blast_database10/groupAlignments_SeqCapAndTargetRef_consensus_InputSeqsMax100.fasta"

GROUPSEQSDIR=$BLASTDBDIR"/supergroupseqs"     # output directory; will include subdirectories
SAMPLESEQSDIR="$BLASTDBDIR""/tempseqs"        # sequences that were used to build the blast databases generated by 03-temp.sh; referred to as $TEMPDIR in some bash scripts
SAMPLESKEY=$BLASTDBDIR"/allsamples_key.txt"
ALIASOUTPATH=$GROUPSEQSDIR/"alias_GenomesSeqsInSupergroups.txt"

# Path to this bash script
SHij="/panfs/pfs.local/home/j926w878/work/conda/snakecap/test13.sh"

[ ! -d "$GROUPSEQSDIR" ] && mkdir "$GROUPSEQSDIR"

if [ $MODE = "A" ] ; then

   ALIAS=$(awk -F '\t' 'NR>1{{print $1"\t"$1"_"$2}}' "$SUPERGROUPMATPATH")
   
   #if [ "$MODE"="A" ] ; then
      SAMPLENAMES=$(awk -F '\t' -v var='sample'  'NR>1{if($1 ~ var) {print $1}}' "$SUPERGROUPMATPATH" | awk '{gsub("_.+","")}1' | sort | uniq | sort -V )
      NUMSAMPLES=$(echo $SAMPLENAMES | wc -w)
      echo "$NUMSAMPLES"
      #for VARi in $(seq 3 "$NUMSAMPLES") ; do
         #  sbatch --nodes=1 --ntasks-per-node=4 --time=6:00:00 --partition=sixhour "$SHij" "B" "$BLASTDBDIR" "$SUPERGROUPMATPATH" "$EDGEMATPATH" "$CONSENSUSPATH" "$VARi"
      #done
   #fi
   
   #if [ "$MODE"="B" ] ; then
   for VARi in $(seq 1 "$NUMSAMPLES") ; do
      SAMPLEi=$(echo $SAMPLENAMES | awk -v i="$VARi" '{print $i}')
      echo $SAMPLEi
      OGNAMEi=$(awk -F ' ' -v var="$SAMPLEi" '{if($1==var){print $2}}' "$SAMPLESKEY")
      SEQNAMESi=$(awk -F '\t' -v var=$SAMPLEi"_" 'NR>1{if($1~var){print $1}}' "$SUPERGROUPMATPATH")
      SEQSPATHi=$(find "$SAMPLESEQSDIR" -name *"$OGNAMEi"*)
      # Extract sequences from the SAMPLEi genome fasta with names matching values in $ALIAS
      SEQSi=$(seqkit grep -n -f <(echo "$SEQNAMESi") "$SEQSPATHi")
      # Rename the extracted sequences from *ALIAS_OLD* to ALIAS_NEW
      TEMPFASTA=$GROUPSEQSDIR"/"$SAMPLEi".fa"
      seqkit replace -p '(.+)$' -r '{kv}' -k <(echo "$ALIAS") <(echo "$SEQSi") > $TEMPFASTA
      # For a particular genome sample, split sequences into separate fasta files by supergroup
      seqkit split "$TEMPFASTA" -i --id-regexp "^.+[0-9]_[0-9]+_([\w]+)$" -2 --quiet --force -w 0
   done
   #   rm $TEMPFASTA
   #fi

   # same as above, but for the ConsensusSeqCapTargRef sequence(s)
   SAMPLENAMES_CON=$(awk -F '\t' -v var='group'  'NR>1{if($1 ~ var) {print $1}}' "$SUPERGROUPMATPATH" | awk '{gsub("_.+","")}1' | sort | uniq | sort -V )
   SEQNAMES_CON=$(awk -F '\t' -v var='group' 'NR>1{if($1~var){print $1}}' "$SUPERGROUPMATPATH")
   SEQS_CON=$(seqkit grep -n -f <(echo "$SEQNAMES_CON") "$CONSENSUSPATH")
   TEMPFASTA=$GROUPSEQSDIR"/SeqCapTargRefConsensus.fa"
   # include supergroup in sequence name
   seqkit replace -p '(.+)$' -r '{kv}' -k <(echo "$ALIAS") <(echo "$SEQS_CON") > $TEMPFASTA
   # split into supergroups (creates a directory for each SG)
   seqkit split "$TEMPFASTA" -i --id-regexp "^.+[0-9]_[0-9]+_([\w]+)$" -2 --quiet -w 0
   #remove $TEMPFASTA
   rm $TEMPFASTA
fi

###############
# Merge all genome sample fasta files of the same group.
###############

#OUTDIR="/panfs/pfs.local/home/j926w878/scratch/scratch_v3/SequenceCapture/SnakeCap_AllSamples/blast_database10/supergroupseqs/supergroups_fasta"
OUTDIR=$GROUPSEQSDIR"/supergroups_fasta_SeqCapTargRefConsensus_GenomesUntrimmed"

[ ! -d "$OUTDIR" ] && mkdir $OUTDIR

PAR=100
if [ $MODE = "C" ] ; then
  GROUPNAMES=$(awk -F '\t' 'NR>1{print $2}' $SUPERGROUPMATPATH | sort | uniq | sort -V)
  # For each group, gather all sequences of the group and put them in a fasta file.
  NGROUPS=$(echo $GROUPNAMES | wc -w)
  touch $GROUPSEQSDIR"/groupsets.txt"
  for VARj in $(seq 1 $PAR $NGROUPS) ; do
    sbatch --nodes=1 --partition=sixhour $SHij $BLASTDBDIR $SUPERGROUPMATPATH "D" $VARj
  done
fi

if [ $MODE = "D" ] ; then
  VARj=$4
  VARj2=$(($VARj+$PAR-1))
  [ $VARj2 -ge $NGROUPS ] && VARj2=$NGROUPS
  SETj=$(echo $GROUPNAMES | cut $"-f""$VARj""-""$VARj2" -d" ")
  for VARi in $(seq 1 $PAR) ; do
    GROUPNAMEij=$(echo $SETj | awk -v i=$VARi '{print $i}')
    cat $(find $GROUPSEQSDIR -regex '.*[^/]*_'${GROUPNAMEij}'.fasta') > $OUTDIR"/"$GROUPNAMEij".fasta" &
  done
  wait
fi

# Now you can run MAFFT on each supergroup fasta containing sequences from genome samples and the supergroup-associated ConsensusSeqCapTargRef(s)




